### ISSUE TITLE
Debug assertion: "Found year with more than one excess day!" with IslamicObservational

### ISSUE THREAD

**Issue by anba** [unicode-org/icu4x#5068]
The following test case asserts with:
```
thread 'main' panicked at /home/andre/.cargo/registry/src/index.crates.io-6f17d22bba15001f/calendrical_calculations-0.1.1/src/islamic.rs:106:13:
assertion `left == right` failed: (ObservationalIslamic) Found year -6149 AH with more than one excess day!
  left: 2
 right: 1
```

Test case:
```rs
use icu::calendar::types::{Era, MonthCode};
use icu::calendar::Calendar;
use icu::calendar::islamic::IslamicObservational;

use std::str::FromStr;

fn main() {
  let cal = IslamicObservational::new_always_calculating();

  let era = Era::from_str("ah").unwrap();
  let month_code = MonthCode::from_str("M01").unwrap();

  let dt = cal.date_from_codes(era, -6149, month_code, 31);
}
```

**Comment by sffc** [comment]
The direction I want to move is to serve pre-computed Islamic years for all positive years from year 0 to something like 1000 years into the future, and then use an algorithmic approximation for years outside that range.

CC @roozbeh

**Comment by roozbehp** [comment]
> The direction I want to move is to serve pre-computed Islamic years for all positive years from year 0 to something like 1000 years into the future, and then use an algorithmic approximation for years outside that range.

Makes sense for all practical purposes, but like for anything pre-computed, it then becomes a data compression problem, with its own interesting issues.

Another issue to watch for is making sure you don't lose or gain a day or two at the border of the computation when you are switching between the two different algorithms.

In the specific case of Islamic calendars, dates occasionally go back up to about 100 years before Hijra, but rarely beyond that. So you probably want a range between -100 and 2500 AH.

**Comment by sffc** [comment]
> In the specific case of Islamic calendars, dates occasionally go back up to about 100 years before Hijra, but rarely beyond that. So you probably want a range between -100 and 2500 AH.

Hmm, how do you render -100? Is it rendered as "-100 AH" or "101 BH"?

**Comment by roozbehp** [comment]
I believe -100 would be 101 BH.

**Comment by Manishearth** [comment]
@roozbehp Is this a guess or based on something you've seen? When I was researching this I had a hard time seeing if the islamic calendars had a "backwards" era like Gregorian, and have not successfully found sources talking about this. If you have sources or at least experience seeing this, that would definitely clarify stuff for us!

**Comment by Manishearth** [comment]
BH with a backwards era (starting at 1) does sound quite plausible. I'm just worried at how hard it was for me to find sources. That said, I was _mostly_ looking at English materials, and a limited amount of Arabic ones.

**Comment by roozbehp** [comment]
> [@roozbehp](https://github.com/roozbehp) Is this a guess or based on something you've seen? When I was researching this I had a hard time seeing if the islamic calendars had a "backwards" era like Gregorian, and have not successfully found sources talking about this. If you have sources or at least experience seeing this (I know it's hard to remember sources for this type of thing), that would definitely clarify stuff for us!

It's my life experience growing up in a Muslim country where they taught History of Islam in mandatory history courses in both school and college. The events of the life of Muhammad (which are partly before Hijra) and the context of the rise of Islam is taught, and his immediate ancestors are mentioned. Here's a Wikipedia article about his grandfather, for example, which uses 127 BH to 102 BH as the years of his life: https://ar.wikipedia.org/wiki/%D9%87%D8%A7%D8%B4%D9%85_%D8%A8%D9%86_%D8%B9%D8%A8%D8%AF_%D9%85%D9%86%D8%A7%D9%81

There is not much interest in the history of the Arabian peninsula going back much beyond that, and that history is typically reported in the Julian calendar.

**Comment by Manishearth** [comment]
Perfect, thank you! Your experience of ground truth is really helpful here.

**Comment by sffc** [comment]
I made a CLDR issue: https://unicode-org.atlassian.net/browse/CLDR-18464

**Comment by Manishearth** [comment]
Fixed in #6876
